src/lib/rbtree.zig

```diff
@@ -0,0 +1,973 @@
+                        //          /   \                  /   \
+                        //       P(R)   U(B)  ------>   C(R)   G(R)
+                        //      /                                \
+                        //   C(R)                                U(B)
+                        //
+                        // If uncle is black, grandparent has to be black, as parent is red.
+                        // If grandparent was red, both of its children would have to be black,
+                        // but parent is red.
+                        // Grandparent color is updated on line 313. Black height on path to B has not changed
+                        // (before: grandparent black and uncle black, now P black and U black)
+                        self.rotate(grandparent, 1 - dir);
+                        parent.color = Color.Black;
+                    } else {
+                        //          G(B)                  G(B)               C(B)
+                        //        /     \                /   \              /   \
+                        //      P(R)   U(B)   ---->    C(R)  U(B)   --->  P(R)  G(R)
+                        //        \                   /                          \
+                        //         C(R)             P(R)                         U(B)
+                        //
+                        // Final recoloring of grandparent is done on line 313 as well.
+                        // Black height on path to U has not changed (before: G and U black, after: C and U black)
+                        // On the old track in P direction, nothing has changed as well
+                        self.rotate(parent, 1 - dir);
+                        self.rotate(grandparent, dir);
+                        current.color = Color.Black;
+                    }
+                    grandparent.color = Color.Red;
+                    break;
+                } else {
+                    //           G(B)                  G(R) <- potential double red fix needed
+                    //         /     \               /     \
+                    //       P(R)   U(R)   ---->   P(B)    U(B)
+                    //      /                     /
+                    //   C(R)                   C(R)
+                    //
+                    // The solution for the case in which uncle is red is to "push blackness down".
+                    // We recolor parent and uncle to black and grandparent to red.
+                    // It is easy to verify that black heights for all nodes have not changed.
+                    // The only problem we have encountered is that grandparent's parent is red.
+                    // If that is the case, we can have double red again. As such, we continue
+                    // fixing by setting `current` to grandparent
+
+                    // If uncle is red, it is not null
+                    std.debug.assert(uncle != null);
+                    const uncle_nonnull = uncle.?;
```

Comment:
From this, the case where `uncle` is `null` is UB/unreachable, so the assert above has no effect when you write `uncle.?`

Reply:
I initially thought that asserting non null explicitely was a good idea, but then I realized that it is dumb. I deleted most of the asserts like this, but apparently forgot some.